proposals/open-ended-ranges.md

```diff
@@ -0,0 +1,260 @@
+`OpenEndRange<T>`, we want to implement that interface in the existing iterable ranges: `IntRange`, `LongRange`, `CharRange`,
+`UIntRange`, `ULongRange`. So they will be implementing both `ClosedRange<T>` and `OpenEndRange<T>` interfaces simultaneously.
+
+```kotlin
+class IntRange : IntProgression(...), ClosedRange<Int>, OpenEndRange<Int> {
+    override val start: Int
+    override val endInclusive: Int
+    override val endExclusive: Int
+}
+```
+
+There's a subtlety in implementing `endExclusive` property in such ranges: usually it returns `endInclusive + 1`, but 
+there can be such ranges where `endInclusive` is already the maximum value of the range type, and so adding one to it
+would overflow.
+
+We decided that in such cases the reasonable behavior would be to throw an exception from the `endExclusive` property 
+getter. The possibility of that will be documented in the base interface, `OpenEndRange`, and additionally the implementation
+of that property will be deprecated in the existing concrete range classes.
+
+### rangeUntil operators for the standard types
+
+`rangeUntil` operators will be provided for the same types and their combinations that currently have `rangeTo` operator defined.
+For the purposes of prototype, we provide them as extension functions, but for consistency we plan to make them members 
+later, before stabilizing the open-ended ranges API.
+
+### Generic open-ended ranges of comparable values
+
+Similar to closed ranges, there will be a function instantiating an open-ended range from any two values of a comparable type:
+
+```kotlin
+operator fun <T : Comparable<T>> T.rangeUntil(that: T): OpenEndRange<T>
+```
+
+### Specialized open-ended ranges of floating point numbers
+
+There also will be two static specializations of `rangeUntil` operator for `Double` and `Float` types of arguments.
+They are special in how they compare values of their bounds with the value passed to `contains` and between themselves,
+so that a range where either bound is NaN is empty, and the `NaN` value is not contained in any range.  
+
+```kotlin
+operator fun Double.rangeUntil(that: Double): OpenEndRange<Double>
+operator fun Float.rangeUntil(that: Float): OpenEndRange<Float>
+```
+
+### Equality of open-ended ranges
+
+Similar to closed ranges, the `OpenEndRange` interface does not specify contract for `equals`/`hashCode` implementations,
+however its concrete implementations can do that. For example, an open-ended range of double values equal to another such range
+when bounds are respectively equal to each other, or to any empty range of doubles, when it is empty itself.
+
+Also, as a consequence of implementing both `OpenEndRange` and `ClosedRange` in concrete range implementations for standard discrete types,
+an open-ended range is equal to the closed range with the same `start` value and `endExlcusive` equal to `endInclusive + 1`:
```

Comment:
```suggestion
an open-ended range is equal to the closed range with the same `start` value and `endExclusive` equal to `endInclusive + 1`:
```

Reply:
One was already fixed by Jake, now this is the last one! Thanks